File: stracture_build/ios/Runner/Pdftron-ios/Utils/PDFTronUtils.swift
//
//  PDFTronUtils.swift
//  Runner
//
//  Created by Maulik Kundaliya on 24/08/22.
//

import Foundation
import pdftron_flutter

@objc class PDFTronUtils : NSObject{
    @objc class func getDocumentController(pdfTronPlugin:PdftronFlutterPlugin) -> PTDocumentController{
        return PdftronFlutterPlugin.pt_getSelectedDocumentController(pdfTronPlugin.tabbedDocumentViewController)
    }
    @objc class func deleteAllAnnotation
    (mPDFViewCtrl:PTPDFViewCtrl, annotType:PTAnnotType){
        var shouldUnlockRead = false
        mPDFViewCtrl.docLockRead()
        shouldUnlockRead = true
        
        let pageCount = mPDFViewCtrl.pageCount
        for i in 1...pageCount{
            let page = mPDFViewCtrl.getDoc()?.getPage(UInt32(i))
            let annotations = mPDFViewCtrl.getAnnotationsOnPage(i)
            for annot in annotations{
                if(annot.isValid()){
                    if(annot.getType() == annotType){
                        page?.annotRemove(with: annot)
                    }
                }
            }
        }
        mPDFViewCtrl.update()
        if (shouldUnlockRead) {
            mPDFViewCtrl.docUnlockRead()
        }
        
    }
    @objc class func deleteAllAnnotationsExcludingLocation(mPDFViewCtrl:PTPDFViewCtrl){
        var shouldUnlockRead = false
        mPDFViewCtrl.docLockRead()
        shouldUnlockRead = true
        
        let pageCount = mPDFViewCtrl.pageCount
        for i in 1...pageCount{
            let page = mPDFViewCtrl.getDoc()?.getPage(UInt32(i))
            let annotations = mPDFViewCtrl.getAnnotationsOnPage(i)
            for annot in annotations{
                if(annot.isValid()){
                    if(!isLocationAnnot(annot: annot)){
                        page?.annotRemove(with: annot)
                    }
                }
            }
        }
        mPDFViewCtrl.update()
        if (shouldUnlockRead) {
            mPDFViewCtrl.docUnlockRead()
        }
        
    }
    @objc class func requestResetRenderingMap(pdfCrtl : PTPDFViewCtrl){
        pdfCrtl.docLock(true)
        pdfCrtl.pageViewMode = e_trn_fit_page
        pdfCrtl.update()
        pdfCrtl.docUnlock()
    }
    @objc class func highlightLocation(mPDFViewCtrl:PTPDFViewCtrl, annotationId:String, isShowRect:Bool){
        //        var isLocationAnnot = true
        var rect : PTPDFRect?
        mPDFViewCtrl.docLockRead()
        var shouldUnlockRead = true
        let pageCount = mPDFViewCtrl.pageCount
        for i in 1...pageCount{
            
            let annotations = mPDFViewCtrl.getAnnotationsOnPage(i)
            for annot in annotations{
                if(annot.isValid() && (annot.getType() != e_ptStamp) && (annot.getSDFObj() != nil) && !annotationId.isEmpty){
                    let obj : PTObj = annot.getSDFObj()
                    let dictIterator : PTObj? = obj.find("NM")
                    if(dictIterator != nil){
                        let isSelectedAnnot = (dictIterator!.getAsPDFText() == annotationId)
                        self.highlightSelectedAnnotation(mPDFViewCtrl: mPDFViewCtrl, annot: annot, colorPt: PTColorPt(x: 0.28, y: 0.76, z: 0.77, w: 0.09), pageNum: Int(i), isShowRect: isShowRect, isCalibratedLocation: isSelectedAnnot)
                        if(isSelectedAnnot && isShowRect){
                            rect = annot.getRect()
                            mPDFViewCtrl.showRect(Int32(i), rect: rect!)
                            mPDFViewCtrl.zoom = mPDFViewCtrl.zoom * 0.8
                        }
                    }
                }
            }
        }
        mPDFViewCtrl.update()
        if (shouldUnlockRead) {
            mPDFViewCtrl.docUnlockRead()
            shouldUnlockRead = false
        }
        if(rect != nil && isShowRect){
            mPDFViewCtrl.showRect(mPDFViewCtrl.currentPage, rect: rect!)
            mPDFViewCtrl.zoom = mPDFViewCtrl.zoom * 0.8
        }
    }
    @objc class func highlightSelectedAnnotation(
        mPDFViewCtrl: PTPDFViewCtrl,
        annot: PTAnnot,
        colorPt: PTColorPt?,
        pageNum: Int,
        isShowRect: Bool = true,
        isCalibratedLocation: Bool = true
    ){
        let mk = PTMarkup(ann: annot)
        if colorPt != nil{
            mk?.setInteriorColor(colorPt, compNum: 3)
        }
        //(colorPt, 3)
        mk?.setOpacity((isCalibratedLocation) ? 0.27 : 0.1)
        mk?.refreshAppearance()
        annot.setColor(colorPt, numcomp: 1)
        if (isShowRect && isCalibratedLocation && annot.getRect() != nil && colorPt != nil) {
            mPDFViewCtrl.showRect(Int32(pageNum), rect: annot.getRect())
            mPDFViewCtrl.zoom = mPDFViewCtrl.zoom * 0.8
//             if (mPDFViewCtrl.zoom > 4){
//                 mPDFViewCtrl.zoom = 4.0
//             }
        }
        annot.refreshAppearance()
    }
    @objc class func removeAllHighlightedAnnotation(mPDFViewCtrl: PTPDFViewCtrl){
        var rect : PTPDFRect?
        mPDFViewCtrl.docLockRead()
        var shouldUnlockRead = true
        let pageCount = mPDFViewCtrl.pageCount
        for i in 1...pageCount{
            
            let annotations = mPDFViewCtrl.getAnnotationsOnPage(i)
            for annot in annotations{
                if(annot.isValid() && (annot.getType() != e_ptStamp) && (annot.getSDFObj() != nil)){
                    let obj : PTObj = annot.getSDFObj()
                    let dictIterator : PTObj? = obj.find("NM")
                    if(dictIterator != nil){
                        self.highlightSelectedAnnotation(mPDFViewCtrl: mPDFViewCtrl, annot: annot, colorPt: PTColorPt(x: 0.28, y: 0.76, z: 0.77, w: 0.09), pageNum: Int(i), isShowRect: false, isCalibratedLocation: false)
//                        if(isSelectedAnnot && isShowRect){
//                            rect = annot.getRect()
//                            mPDFViewCtrl.showRect(Int32(i), rect: rect!)
//                            mPDFViewCtrl.zoom = mPDFViewCtrl.zoom * 0.8
//                        }
                    }
                }
            }
        }
        mPDFViewCtrl.update()
        if (shouldUnlockRead) {
            mPDFViewCtrl.docUnlockRead()
            shouldUnlockRead = false
        }
//        if(rect != nil && isShowRect){
//            mPDFViewCtrl.showRect(mPDFViewCtrl.currentPage, rect: rect!)
//            mPDFViewCtrl.zoom = mPDFViewCtrl.zoom * 0.8
//        }
    }
    @objc class func isLocationAnnot(annot: PTAnnot) -> Bool {
        return annot.isValid() && (((annot.getType().rawValue)) >= 4 && ((annot.getType().rawValue)) <= 7)
    }
    @objc class func getAnnotationIdAt(pdfCtrl : PTPDFViewCtrl,x:Double,y:Double) -> String{
        let annot = pdfCtrl.getAnnotationAt(Int32(x), y: Int32(y), distanceThreshold: 0.5, minimumLineWeight: 0.5)
        if (annot != nil && annot?.getSDFObj() != nil && (annot?.getSDFObj().find("NM").getAsPDFText()) != nil)
        {
            return (annot?.getSDFObj().find("NM").getAsPDFText())!
        }
        return ""
        
    }
    @objc class func getAnnotById(mPdfControl: PTPDFViewCtrl, annotationId: String?) -> PTAnnot? {
        var annot: PTAnnot?
        mPdfControl.docLock(true)
        for i in 1...mPdfControl.pageCount{
            let annotations = mPdfControl.getAnnotationsOnPage(Int32(i))
            for tmpAnnot in annotations{
                let obj : PTObj = tmpAnnot.getSDFObj()
                let dictIterator : PTObj? = obj.find("NM")
                if(dictIterator != nil){
                    let isSelectedAnnot = (dictIterator!.getAsPDFText() == annotationId)
                    if(isSelectedAnnot){
                        annot = tmpAnnot
                        break
                    }
                }
            }
            if(annot != nil){
                break
            }
        }
        mPdfControl.update()
        mPdfControl.docUnlock()
        return annot
    }
    @objc class func scrollToCenter(mPDFViewCtrl: PTPDFViewCtrl, annotation: PTAnnot, currentPageNumber: Int) {
        let screenRect = PDFTronUtils.getOldAnnotScreenPosition(mPDFViewCtrl: mPDFViewCtrl, annotation: annotation, currentPageNumber: currentPageNumber)
        
        //show as center of screen
        let annotCentreX: Double =
        (screenRect!.x1 + (screenRect!.x1 + screenRect!.width())) / 2
        let annotCentreY: Double =
        (screenRect!.y1 + (screenRect!.y1 + screenRect!.height())) / 2
        let screenCentreX: Double = Double(mPDFViewCtrl.frame.size.width / 2)
        let screenCentreY: Double = Double(mPDFViewCtrl.frame.size.height / 2)
        let finalCentreX = annotCentreX - screenCentreX
        let finalCentreY = annotCentreY - screenCentreY
        scrollToDocPosition(mPdfControl: mPDFViewCtrl, horizontalPosition: Int(finalCentreX), verticalPosition: Int(finalCentreY), page: currentPageNumber)
        mPDFViewCtrl.update()
    }
    private class func getOldAnnotScreenPosition(mPDFViewCtrl: PTPDFViewCtrl, annotation: PTAnnot, currentPageNumber: Int) -> PTPDFRect?{
            if(annotation.isValid()){
                let oldUpdateRect = mPDFViewCtrl.getScreenRect(for: annotation, page_num: Int32(currentPageNumber))
                oldUpdateRect?.normalize()
                return oldUpdateRect
            }
            return nil
        }
    private class func scrollToDocPosition(
        mPdfControl: PTPDFViewCtrl,
        horizontalPosition: Int,
        verticalPosition: Int,
        page: Int
    ) {
        let pts = PTPDFPoint(px: Double(horizontalPosition), py: Double(verticalPosition))
        let screenPt = mPdfControl.convScreenPt(toPagePt: pts!, page_num: Int32(page))
        let pagePt = mPdfControl.convPagePt(toScreenPt: screenPt, page_num: Int32(page))
        
        mPdfControl.setHScrollPos(pagePt.x + mPdfControl.getHScrollPos())
        mPdfControl.setVScrollPos(pagePt.y + mPdfControl.getVScrollPos())
    }
    @objc class func deleteImageView(mPDFViewCtrl:PTPDFViewCtrl, annotId: Double)
    {
        for subview in mPDFViewCtrl.subviews{
            if subview.accessibilityIdentifier == String(annotId-1) {
                subview.removeFromSuperview()
            }
        }
        
    }
    @objc class func polygonAnnotation(mPDFViewCtrl:PTPDFViewCtrl, x1:Double,y1:Double,locationAngle:Double,annotId:Double){
        do {
            try PTPDFNet.catchException
            {
                let doc: PTPDFDoc? = mPDFViewCtrl.getDoc()

                if(doc != nil){
                    let page: PTPage? = doc!.getPage(UInt32(mPDFViewCtrl.currentPage))
                    let width = page?.getWidth(PTBox(UInt32(getpagesize())))
                    let height = page?.getHeight(PTBox(UInt32(getpagesize())))
                    var rotation = page?.getRotation().rawValue
                           
                    var x = 0.0
                    var y = 0.0

                    let widthMinusY1 = y1
                    var angleOfRotation = 0.0;

                    
                    if(rotation! == 0){
                        x = x1 / 2
                        y = widthMinusY1 / 2
                        angleOfRotation = 0.0
                    }

                    if(rotation! == 1){
                        x = height! - widthMinusY1 / 2
                        y = x1 / 2
                        angleOfRotation = .pi / 2
                    }

                    if(rotation! == 2){
                        x = width! - (x1 / 2)
                        y = height! - (widthMinusY1 / 2)
                        angleOfRotation = .pi
                    }

                    if(rotation! == 3){
                        x = widthMinusY1 / 2
                        y = (height! - x1) / 2
                        angleOfRotation = .pi / 2 
                    }

                    
                   //let angleVal = angleOfRotation * (CGFloat(180) / (.pi) )//(angleOfRotation * 360) / (3.14159265*2)
                    let degree = (locationAngle) * (.pi/180)
                    print(degree)
                    let pointTemp = mPDFViewCtrl.convCanvasPt(toScreenPt: PTPDFPoint(px: x, py: y ))
                    print(mPDFViewCtrl.zoom)
                    let userPositionMarkup =  UIView(frame: CGRect(x: pointTemp.x-15, y: pointTemp.y-15, width: CGFloat(30.0), height: CGFloat(30.0)))
                    userPositionMarkup.clipsToBounds = false
                    userPositionMarkup.accessibilityIdentifier = String(annotId)
                    let imgvPin = UIImageView(frame: CGRect(x: 0.0, y: 0.0, width: userPositionMarkup.frame.size.width, height: userPositionMarkup.frame.size.height))
                    imgvPin.image = UIImage(named: "user_location")

                    imgvPin.transform = CGAffineTransformMakeRotation(-degree - (angleOfRotation));
                    userPositionMarkup.addSubview(imgvPin)
                    mPDFViewCtrl.addSubview(userPositionMarkup)
                    if(annotId > 0.0){
                        deleteImageView(mPDFViewCtrl: mPDFViewCtrl, annotId: annotId)
                    }
                }
            }
        } catch
        {
            print("Caught exception")
        }
        
    }
    @objc class func deleteAnnotationsOnPage(mPDFViewCtrl:PTPDFViewCtrl,  annotType:PTAnnotType) {
        var shouldUnlockRead = false
        mPDFViewCtrl.docLockRead()
        shouldUnlockRead = true
        
        let page = mPDFViewCtrl.getDoc()?.getPage(UInt32(1))
        let annotations = mPDFViewCtrl.getAnnotationsOnPage(1)
        if(annotations.count>2)
        {
            for annot in annotations{
                if(annot.isValid()){
                    if(annot.getType() == annotType){
                        page?.annotRemove(with: annot)
                    }
                }
            }
            page?.annotRemove(with: annotations.first)
            
        }
        mPDFViewCtrl.update()
        if (shouldUnlockRead) {
            mPDFViewCtrl.docUnlockRead()
        }
    }
}

File: stracture_build/android/app/src/main/kotlin/com/asite/field/pdftron/utils/PdftronUtils.kt
package com.asite.field.pdftron.utils


import android.content.Context
import android.graphics.Color
import android.graphics.PorterDuff
import android.text.TextUtils
import android.util.DisplayMetrics
import android.view.View
import android.view.animation.Animation
import android.view.animation.ScaleAnimation
import android.view.inputmethod.InputMethodManager
import android.widget.ImageView
import androidx.core.content.res.ResourcesCompat
import com.asite.field.pdftron.listener.SiteOperationListener
import com.asite.field.pdftron.view.BottomCenterAnchoredCustomLayout
import com.asite.field.pdftron.view.BottomCenterAnchoredCustomLayoutLocation
import com.asite.field.R
import com.asite.field.pdftron.constant.PdftronConstant
import com.pdftron.common.PDFNetException
import com.pdftron.pdf.Annot
import com.pdftron.pdf.ColorPt
import com.pdftron.pdf.PDFViewCtrl
import com.pdftron.pdf.Rect
import com.pdftron.pdf.annots.Markup
import com.pdftron.pdf.utils.AnnotUtils
import com.pdftron.pdf.utils.Utils
import java.util.*
import kotlin.math.roundToInt
import kotlin.math.round
import kotlin.math.abs
import android.widget.RelativeLayout


class PdftronUtils {

    companion object {
        const val viewTypeId = "asite/pdftron_flutter/documentview"

        @JvmStatic
        fun drawPinAsView(
            pdfViewCtrl: PDFViewCtrl,
            x: Double,
            y: Double,
            mDownPageNum: Int,
            colorCode: String,
            formTypeCode: String,
            isHighlight: Boolean,
            annotationId: String,
            siteOperationListener: SiteOperationListener?
        ) {
            removePinAnnotView(pdfViewCtrl, annotationId)
            val context = pdfViewCtrl.context
            val overlay = BottomCenterAnchoredCustomLayout(
                context, pdfViewCtrl, x, y, mDownPageNum
            )

            pdfViewCtrl.addView(overlay)
            val pinSize: Int = context.dpToPx(20f)
            val image = ImageView(context)
            image.minimumHeight = pinSize * 2
            image.minimumWidth = pinSize * 2
            image.scaleType = ImageView.ScaleType.FIT_XY
            image.setImageDrawable(
                ResourcesCompat.getDrawable(
                    context.resources, R.drawable.marker, null
                )
            )

            image.setColorFilter(Color.parseColor(colorCode), PorterDuff.Mode.SRC_IN)
            overlay.addView(image)
            overlay.annotationId = annotationId
            val screenPoint = pdfViewCtrl.convPagePtToScreenPt(x, y, mDownPageNum)
            overlay.setScreenRect(
                screenPoint[0] - pinSize / 2,
                screenPoint[1] - pinSize,
                screenPoint[0] + pinSize / 2,
                screenPoint[1],
                mDownPageNum
            )
            overlay.setBackgroundColor(Color.rgb(255, 0, 0))
            overlay.setZoomWithParent(false)
            scaleView(overlay)
            overlay.setOnClickListener {
                siteOperationListener?.onAnnotationSelectedListener(
                    (it as BottomCenterAnchoredCustomLayout).annotationId.toString(), false
                )
            }
        }

        private fun removePinAnnotView(mPdfControl: PDFViewCtrl, annotationId: String) {
            try {
                if (!TextUtils.isEmpty(annotationId)) {
                    val createdPin: BottomCenterAnchoredCustomLayout? =
                        mPdfControl.findViewWithTag(annotationId)
                    if (createdPin != null) {
                        mPdfControl.removeView(createdPin)
                    }
                }
            } catch (ex: Exception) {
                ex.printStackTrace()
            }
        }

        private fun removeLocationPinAnnotView(mPdfControl: PDFViewCtrl, annotationId: String) {
            try {
                if (!TextUtils.isEmpty(annotationId)) {
                    val createdPin: BottomCenterAnchoredCustomLayoutLocation? =
                        mPdfControl.findViewWithTag(annotationId)
                    if (createdPin != null) {
                        mPdfControl.removeView(createdPin)
                    }
                }
            } catch (ex: Exception) {
                ex.printStackTrace()
            }
        }


        @JvmStatic
        fun scaleView(v: View, startScale: Float = 0f, endScale: Float = 1f) {
            val anim: Animation = ScaleAnimation(
                1f, 1f,  // Start and end values for the X axis scaling
                startScale, endScale,  // Start and end values for the Y axis scaling
                Animation.RELATIVE_TO_SELF, 0f,  // Pivot point of X scaling
                Animation.RELATIVE_TO_SELF, 1f
            ) // Pivot point of Y scaling
            anim.fillAfter = true // Needed to keep the result of the animation
            anim.duration = 1000
            v.startAnimation(anim)
        }

        @JvmStatic
        fun highlightSelectedAnnotation(
            mPDFViewCtrl: PDFViewCtrl,
            annot: Annot,
            colorPt: ColorPt,
            pageNum: Int,
            isShowRect: Boolean,
            isCalibratedLocation: Boolean = true
        ) {
            try {
                mPDFViewCtrl.docLock(true) {
                    val mk = Markup(annot)
                    //set Border Color
                    annot.setColor(Utils.color2ColorPt(Color.TRANSPARENT), 3)
                    mk.setInteriorColor(colorPt, 3)
                    mk.opacity = if (isCalibratedLocation) 0.27 else 0.1
                    if (isShowRect && isCalibratedLocation && annot.rect != null) {
                        mPDFViewCtrl.showRect(pageNum, annot.rect)
                        mPDFViewCtrl.zoom = (mPDFViewCtrl.zoom * 0.8)

                    }
                    annot.refreshAppearance()
                }
            } catch (ex: Exception) {
                ex.printStackTrace()
            }
        }

        @JvmStatic
        fun deleteAllAnnotations(mPDFViewCtrl: PDFViewCtrl, annotType: Int = -1) {
            try {
                mPDFViewCtrl.docLockRead()
                val shouldUnlockRead = true
                try {
                    for (i in 1..mPDFViewCtrl.doc.pageCount) {
                        val page = mPDFViewCtrl.doc!!.getPage(i)
                        var annotCount = page.numAnnots
                        while (annotCount > 0) {
                            if (page.getAnnot(annotCount - 1).isValid) {
                                if (annotType == -1) {
                                    page.annotRemove(annotCount - 1)
                                } else if (page.getAnnot(annotCount - 1).type == annotType) {
                                    page.annotRemove(annotCount - 1)
                                }
                                annotCount--
                            }
                        }
                    }
                    mPDFViewCtrl.update(true)

                } catch (e: PDFNetException) {
                    e.printStackTrace()
                } finally {
                    if (shouldUnlockRead) {
                        mPDFViewCtrl.docUnlockRead()
                    }
                }
            } catch (e: Exception) {
            }
        }

        @JvmStatic
        fun deleteAllAnnotationsExcludingLocation(mPDFViewCtrl: PDFViewCtrl) {
            try {
                mPDFViewCtrl.docLockRead()
                val shouldUnlockRead = true
                try {
                    for (i in 1..mPDFViewCtrl.doc.pageCount) {
                        val page = mPDFViewCtrl.doc!!.getPage(i)
                        var annotCount = page.numAnnots
                        while (annotCount > 0) {
                            if (page.getAnnot(annotCount - 1).isValid) {
                                if (!isLocationAnnot(page.getAnnot(annotCount - 1))) {
                                    page.annotRemove(annotCount - 1)
                                }
                                annotCount--
                            }
                        }
                    }
                    mPDFViewCtrl.update(true)

                } catch (e: PDFNetException) {
                    e.printStackTrace()
                } finally {
                    if (shouldUnlockRead) {
                        mPDFViewCtrl.docUnlockRead()
                    }
                }
            } catch (e: Exception) {
            }
        }

        @JvmStatic
        fun highlightLocation(mPdfControl: PDFViewCtrl, annot: Annot) {
            try {
                if (isLocationAnnot(annot) && annot.sdfObj.findObj(
                        "NM"
                    ).asPDFText != null
                ) {
                    mPdfControl.docLock(true) {
                        try {
                            for (i in 1..mPdfControl.doc.pageCount) {
                                for (j in 0 until mPdfControl.doc.getPage(i).numAnnots) {
                                    val tempAnnot = mPdfControl.doc.getPage(i).getAnnot(j)
                                    val isSelectedAnnot =
                                        tempAnnot.sdfObj.findObj("NM").asPDFText == annot.sdfObj.findObj(
                                            "NM"
                                        ).asPDFText
                                    highlightSelectedAnnotation(
                                        mPdfControl,
                                        tempAnnot,
                                        PdftronConstant.highlightColorPt,
                                        i,
                                        isCalibratedLocation = isSelectedAnnot,
                                        isShowRect = isSelectedAnnot
                                    )
                                }
                            }
                            mPdfControl.update()
                        } catch (ex: PDFNetException) {
                            ex.printStackTrace()
                        }
                    }
                }
            } catch (e: PDFNetException) {
                e.printStackTrace()
            }

        }

        @JvmStatic
        fun highlightLocation(
            mPdfControl: PDFViewCtrl, annotationId: String?, isShowRect: Boolean
        ) {
            try {
                mPdfControl.docLock(true) {
                    try {
                        val pageNum: Int = mPdfControl.doc.pageCount
                        for (i in 1..pageNum) {
                            val annotList: ArrayList<Annot> = mPdfControl.getAnnotationsOnPage(i)
                            val annotIterator: Iterator<Annot> = annotList.iterator()
                            while (annotIterator.hasNext()) {
                                val annot = annotIterator.next()
                                if (isLocationAnnot(annot)) {
                                    val isSelectedAnnot = !annot.sdfObj.findObj(
                                        "NM"
                                    ).asPDFText.isNullOrEmpty() && !annotationId.isNullOrEmpty() && annotationId.lowercase() != "null" && annot.sdfObj.findObj(
                                        "NM"
                                    ).asPDFText == annotationId
                                    highlightSelectedAnnotation(
                                        mPdfControl,
                                        annot,
                                        PdftronConstant.highlightColorPt,
                                        i,
                                        isCalibratedLocation = isSelectedAnnot,
                                        isShowRect = isShowRect
                                    )
                                }
                            }

                        }
                        mPdfControl.update()
                    } catch (ex: PDFNetException) {
                        ex.printStackTrace()
                    }
                }
            } catch (e: PDFNetException) {
                e.printStackTrace()
            }
        }

        @JvmStatic
        fun requestResetRenderingPdftron(mPdfControl: PDFViewCtrl) {
            try {
                mPdfControl.docLock(true) {
                    // Set preferred view mode to PageViewMode.FIT_PAGE
                    mPdfControl.preferredViewMode = PDFViewCtrl.PageViewMode.FIT_PAGE
                    // Set PageViewMode.FIT_PAGE to current page view mode
                    mPdfControl.pageViewMode = PDFViewCtrl.PageViewMode.FIT_PAGE
                    mPdfControl.directionalLockEnabled = false
                }
            } catch (ex: PDFNetException) {
                ex.printStackTrace()
            }
        }

        @JvmStatic
        fun isLocationAnnot(annot: Annot): Boolean {
            return annot.isValid && annot.type >= Annot.e_Square && annot.type <= Annot.e_Polyline
        }

        fun getAnnotationIdAt(mPdfControl: PDFViewCtrl, x: Double, y: Double): String? {
            try {
                val annot = mPdfControl.getAnnotationAt(x.toInt(), y.toInt())
                if (annot != null && annot.sdfObj != null && !annot.sdfObj.findObj(
                        "NM"
                    ).asPDFText.isNullOrEmpty()
                ) {
                    return annot.sdfObj.findObj("NM").asPDFText
                }
            } catch (e: Exception) {
            }
            return null
        }

        fun getUniqueAnnotationId(): String? {
            return UUID.randomUUID().toString() + "-" + Calendar.getInstance().timeInMillis
        }

        fun Context.dpToPx(dp: Float): Int {
            val displayMetrics: DisplayMetrics = this.resources.displayMetrics
            return (dp * displayMetrics.density).roundToInt()
        }

        fun setInteractionEnabled(mPdfControl: PDFViewCtrl, interactionEnabled: Boolean) {
            mPdfControl.setInteractionEnabled(interactionEnabled)
        }

        /**
         * Hides soft keyboard.
         *
         * @param context The context
         * @param view    The view
         */
        fun hideSoftKeyboard(context: Context, view: View) {
            try {
                val imm =
                    context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                if (imm != null && imm.isActive) {
                    imm.hideSoftInputFromWindow(view.windowToken, 0)
                }
            } catch (ignored: java.lang.Exception) {
            }
        }

        //This code snippet takes in coordinates of four points in a PDF document and converts them to screen points. It then creates an annotation on the PDF document using these screen points and adds an image to the annotation. The image is rotated based on the slope of
        //This code defines a polygon annotation that takes two screen coordinates and adds it to the PDFViewCtrl. It also sets up an image view with minimum height, maximum width, and scale type of images based on their user location

        @JvmStatic

        fun polygonAnnotation(
            mPDFViewCtrl: PDFViewCtrl,
            x1: Double,
            y1: Double,
            locationAngle: Double,
            annotId: Double
        ) {

            try {

                if (mPDFViewCtrl.doc != null) {

                    val page = mPDFViewCtrl.doc.getPage(1)
                    val rotation = page.getRotation();
                    val width = page.getPageWidth()
                    val height = page.getPageHeight()
                    var rotationAngle = (-locationAngle + 360) % 360

                    var x = 0.0;
                    var y = 0.0;

                    if(rotation == 0){
                        x = x1
                        y = y1
                    }
                    else if(rotation == 1){
                        x = height - y1;
                        y = x1;
                        rotationAngle += 90
                    }
                    else if(rotation == 2){
                        x = width  - x1;
                        y = height - y1;
                        rotationAngle += 180
                    }
                    else if(rotation == 3){
                        x = y1;
                        y = height - x1
                        rotationAngle += 270
                    }

                    val pagePoints = mPDFViewCtrl
                        .convCanvasPtToPagePt(x.toDouble(), y.toDouble(),-1);

                    val context = mPDFViewCtrl.context;
                    val pinSize: Int = context.dpToPx(50f);

                    /* Use PAGE COORDINATES WHEN PASSING DATA */
                    val overlay = BottomCenterAnchoredCustomLayoutLocation(
                        context, mPDFViewCtrl, pagePoints[0], pagePoints[1], 1
                    )

                    mPDFViewCtrl.addView(overlay)
                    val image = ImageView(context)
                    image.scaleType = ImageView.ScaleType.FIT_XY
                    image.setImageDrawable(
                        ResourcesCompat.getDrawable(
                            context.resources, R.drawable.user_location_icon, null
                        )
                    );


                    image.setRotation(rotationAngle.toFloat());
                    image.isClickable = false;
                    image.isFocusable = false;
                    //image.setBackgroundColor(Color.rgb(255, 0, 0))


                    overlay.addView(image);
                    overlay.annotationId = annotId.toString();
                    //overlay.setBackgroundColor(Color.rgb(255, 255, 0))


                    val layoutParams = RelativeLayout.LayoutParams(
                        RelativeLayout.LayoutParams.MATCH_PARENT,
                        RelativeLayout.LayoutParams.MATCH_PARENT
                    )
                    
                    image.layoutParams = layoutParams

                /* USE SCREEN COORDINATES WHEN PASSING RECT */        
                val screenPoint = mPDFViewCtrl.convCanvasPtToScreenPt(x, y, 1)
                    overlay.setScreenRect(
                        screenPoint[0] - pinSize / 2,
                        screenPoint[1] - pinSize / 2,
                        screenPoint[0] + pinSize / 2,
                        screenPoint[1] + pinSize / 2,
                        1
                    );

                  

                    if (annotId > 0) {
                        removeLocationPinAnnotView(mPDFViewCtrl, (annotId - 1.0).toString())
                    }


                }

            } catch (e: PDFNetException) {

                e.printStackTrace()

            }

        }

        @JvmStatic
        fun deleteAnnotationsOnPage(pdfViewCtrl: PDFViewCtrl, annot: Annot?, pageNum: Int) {
            pdfViewCtrl.docLock(true) {
                val page = pdfViewCtrl.doc.getPage(pageNum)
                var annotCount = page.numAnnots
                while (annotCount > 1) {
                    page.annotRemove(0)
                    annotCount = page.numAnnots
                }
                pdfViewCtrl.update(annot, pageNum)
                pdfViewCtrl.update(false)
            }
        }

        fun getAnnotById(mPdfControl: PDFViewCtrl, annotationId: String?): Annot? {
            var annot: Annot? = null
            try {
                mPdfControl.docLock(true) {
                    try {
                        for (i in 1..mPdfControl.doc.pageCount) {
                            for (j in 0 until mPdfControl.doc.getPage(i).numAnnots) {
                                val tempAnnot = mPdfControl.doc.getPage(i).getAnnot(j)
                                if (isLocationAnnot(tempAnnot)) {
                                    val isSelectedAnnot = !tempAnnot.sdfObj.findObj(
                                        "NM"
                                    ).asPDFText.isNullOrEmpty() && !annotationId.isNullOrEmpty() && annotationId.lowercase() != "null" && tempAnnot.sdfObj.findObj(
                                        "NM"
                                    ).asPDFText == annotationId
                                    if (isSelectedAnnot) {
                                        annot = tempAnnot
                                        break
                                    }
                                }
                            }
                            if (annot != null) break
                        }
                        mPdfControl.update()
                    } catch (ex: PDFNetException) {
                        ex.printStackTrace()
                    }
                }
            } catch (e: PDFNetException) {
                e.printStackTrace()
            }
            return annot
        }

        @JvmStatic
        fun scrollToCenter(mPDFViewCtrl: PDFViewCtrl, annotation: Annot, currentPageNumber: Int) {
            try {
                val screenRect: Rect = AnnotUtils.getOldAnnotScreenPosition(
                    mPDFViewCtrl, annotation, currentPageNumber
                )
                //show as center of screen
                val annotCentreX: Double = (screenRect.x1 + (screenRect.x1 + screenRect.width)) / 2
                val annotCentreY: Double = (screenRect.y1 + (screenRect.y1 + screenRect.height)) / 2
                val screenCentreX: Int = mPDFViewCtrl.width / 2
                val screenCentreY: Int = mPDFViewCtrl.height / 2
                val finalCentreX = annotCentreX - screenCentreX
                val finalCentreY = annotCentreY - screenCentreY
                try {
                    scrollToDocPosition(
                        mPDFViewCtrl, finalCentreX.toInt(), finalCentreY.toInt(), currentPageNumber
                    )
                } catch (e: Exception) {
                    e.printStackTrace()
                }
                mPDFViewCtrl.update()
            } catch (e: PDFNetException) {
                e.printStackTrace()
            }
        }

        private fun scrollToDocPosition(
            mPdfControl: PDFViewCtrl, horizontalPosition: Int, verticalPosition: Int, page: Int
        ) {
            val screenPt: DoubleArray = mPdfControl.convScreenPtToPagePt(
                horizontalPosition.toDouble(), verticalPosition.toDouble(), page
            )
            val pagePt: DoubleArray = mPdfControl.convPagePtToScreenPt(
                screenPt[0], screenPt[1], page
            )
            mPdfControl.hScrollPos = pagePt[0].toInt() + mPdfControl.hScrollPos
            mPdfControl.vScrollPos = pagePt[1].toInt() + mPdfControl.vScrollPos
        }
    }
}
